File: 1.single_joint_control/cpp/single_joint_control/src/cmd_publisher.cpp
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "ros2_bridge_msgs/msg/arm_ctrl.hpp"
#include "ros2_bridge_msgs/msg/motor_ctrl.hpp"
#include "ros2_bridge_msgs/msg/robot_state.hpp"

class CmdPublisher : public rclcpp::Node {
 public:
  CmdPublisher()
      : Node("cmd_publisher"), current_pos_(0.0), target_pos_(0.0), stage_(0) {
    publisher_ =
        this->create_publisher<ros2_bridge_msgs::msg::ArmCtrl>("/arm/cmd", 10);

    status_sub_ =
        this->create_subscription<ros2_bridge_msgs::msg::RobotState>(
            "/robot_state", 10,
            std::bind(&CmdPublisher::status_callback, this,
                      std::placeholders::_1));

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(20),
        std::bind(&CmdPublisher::control_callback, this));

    RCLCPP_INFO(this->get_logger(), "CmdPublisher initialized");
  }

 private:
  static constexpr uint16_t kJointId = 21;
  static constexpr double kPosTarget = -1.588;
  static constexpr double kPosThresholdDeg = 5.0;

  // Position mode parameters
  static constexpr uint8_t kPosMode = 0;
  static constexpr double kPosSpd = 0.3;
  static constexpr double kPosCur = 10.0;

  // Force-position hybrid mode parameters
  static constexpr uint8_t kHybridMode = 1;
  static constexpr double kHybridKp = 50.0;
  static constexpr double kHybridKd = 2.0;
  static constexpr double kHybridSpd = 0.0;
  static constexpr double kHybridTor = 0.0;

  void status_callback(
      const ros2_bridge_msgs::msg::RobotState::SharedPtr msg) {
    for (const auto& motor_status : msg->arm.status) {
      if (motor_status.name == kJointId) {
        current_pos_ = motor_status.pos;
      }
    }
  }

  double pos_diff_deg(double target) const {
    return std::abs(current_pos_ - target) * 180.0 / M_PI;
  }

  void control_callback() {
    switch (stage_) {
      case 0:
        run_position_stage(kPosTarget, 1, "Position mode: raise arm");
        break;
      case 1:
        run_position_stage(0.0, 2, "Position mode: lower arm");
        break;
      case 2:
        run_hybrid_stage(kPosTarget, 3, "Hybrid mode: raise arm");
        break;
      case 3:
        run_hybrid_stage(0.0, 4, "Hybrid mode: lower arm");
        break;
      default:
        break;
    }
  }

  void run_position_stage(double target, int next_stage, const char* label) {
    target_pos_ = target;
    double diff = pos_diff_deg(target);
    RCLCPP_INFO(this->get_logger(),
                "[Stage %d] %s: current=%.4f, target=%.4f, diff=%.2f deg",
                stage_, label, current_pos_, target, diff);

    if (diff < kPosThresholdDeg) {
      RCLCPP_INFO(this->get_logger(), "Stage %d completed", stage_);
      stage_ = next_stage;
      if (next_stage == 2) {
        RCLCPP_INFO(this->get_logger(),
                    "--- Switching to force-position hybrid mode ---");
      }
    }
    publish_position_cmd(target);
  }

  void run_hybrid_stage(double target, int next_stage, const char* label) {
    target_pos_ = target;
    double diff = pos_diff_deg(target);
    RCLCPP_INFO(this->get_logger(),
                "[Stage %d] %s: current=%.4f, target=%.4f, diff=%.2f deg",
                stage_, label, current_pos_, target, diff);

    if (diff < kPosThresholdDeg) {
      RCLCPP_INFO(this->get_logger(), "Stage %d completed", stage_);
      stage_ = next_stage;
      if (next_stage == 4) {
        RCLCPP_INFO(this->get_logger(), "All motion stages completed!");
        timer_->cancel();
      }
    }
    publish_hybrid_cmd(target);
  }

  void publish_position_cmd(double pos) {
    ros2_bridge_msgs::msg::ArmCtrl msg;
    msg.header.stamp = this->now();
    msg.header.frame_id = "arm";
    msg.mode = kPosMode;
    msg.label = 0;

    ros2_bridge_msgs::msg::MotorCtrl ctrl;
    ctrl.name = kJointId;
    ctrl.pos = pos;
    ctrl.spd = kPosSpd;
    ctrl.cur = kPosCur;
    msg.ctrl.push_back(ctrl);

    publisher_->publish(msg);
  }

  void publish_hybrid_cmd(double pos) {
    ros2_bridge_msgs::msg::ArmCtrl msg;
    msg.header.stamp = this->now();
    msg.header.frame_id = "arm";
    msg.mode = kHybridMode;
    msg.label = 0;

    ros2_bridge_msgs::msg::MotorCtrl ctrl;
    ctrl.name = kJointId;
    ctrl.kp = kHybridKp;
    ctrl.kd = kHybridKd;
    ctrl.pos = pos;
    ctrl.spd = kHybridSpd;
    ctrl.tor = kHybridTor;
    msg.ctrl.push_back(ctrl);

    publisher_->publish(msg);
  }

  rclcpp::Publisher<ros2_bridge_msgs::msg::ArmCtrl>::SharedPtr publisher_;
  rclcpp::Subscription<ros2_bridge_msgs::msg::RobotState>::SharedPtr
      status_sub_;
  rclcpp::TimerBase::SharedPtr timer_;

  double current_pos_;
  double target_pos_;
  int stage_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CmdPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: 1.single_joint_control/python/single_joint_control/cmd_publisher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Single Joint Control Node (Python version)

This node controls a single joint motor using ROS2 topic publishing.
It demonstrates position mode and force-position hybrid mode control.
"""

import math

import rclpy
from rclpy.node import Node

from ros2_bridge_msgs.msg import ArmCtrl, MotorCtrl, RobotState


class CmdPublisher(Node):
    """ROS2 Node for single joint motor control

    Stages:
      0 - Position mode:              move joint 21 to -1.588 rad
      1 - Position mode:              move joint 21 back to 0.0 rad
      2 - Force-position hybrid mode: move joint 21 to -1.588 rad
      3 - Force-position hybrid mode: move joint 21 back to 0.0 rad
      4 - Completed
    """

    JOINT_ID = 21  # Right Shoulder Pitch
    POS_TARGET = -1.588
    POS_THRESHOLD_DEG = 5.0
    CONTROL_INTERVAL_SEC = 0.02

    # Position mode parameters
    POS_MODE = 0
    POS_SPD = 0.3
    POS_CUR = 10.0

    # Force-position hybrid mode parameters
    HYBRID_MODE = 1
    HYBRID_KP = 50.0
    HYBRID_KD = 2.0
    HYBRID_SPD = 0.0
    HYBRID_TOR = 0.0

    def __init__(self):
        super().__init__('cmd_publisher')

        self.current_pos = 0.0
        self.target_pos = 0.0
        self.stage = 0

        self.publisher = self.create_publisher(ArmCtrl, '/arm/cmd', 10)

        self.status_sub = self.create_subscription(
            RobotState, '/robot_state', self.status_callback, 10
        )

        self.timer = self.create_timer(
            self.CONTROL_INTERVAL_SEC, self.control_callback
        )

        self.get_logger().info('CmdPublisher initialized')

    def status_callback(self, msg):
        """Extract joint 21 position from RobotState"""
        for motor_status in msg.arm.status:
            if motor_status.name == self.JOINT_ID:
                self.current_pos = motor_status.pos

    def control_callback(self):
        if self.stage == 0:
            self._run_position_stage(self.POS_TARGET, next_stage=1,
                                     label='Position mode: raise arm')
        elif self.stage == 1:
            self._run_position_stage(0.0, next_stage=2,
                                     label='Position mode: lower arm')
        elif self.stage == 2:
            self._run_hybrid_stage(self.POS_TARGET, next_stage=3,
                                   label='Hybrid mode: raise arm')
        elif self.stage == 3:
            self._run_hybrid_stage(0.0, next_stage=4,
                                   label='Hybrid mode: lower arm')

    def _pos_diff_deg(self, target):
        return abs(self.current_pos - target) * 180.0 / math.pi

    def _run_position_stage(self, target, next_stage, label):
        self.target_pos = target
        diff = self._pos_diff_deg(target)
        self.get_logger().info(
            f'[Stage {self.stage}] {label}: '
            f'current={self.current_pos:.4f}, target={target:.4f}, '
            f'diff={diff:.2f} deg'
        )
        if diff < self.POS_THRESHOLD_DEG:
            self.get_logger().info(f'Stage {self.stage} completed')
            self.stage = next_stage
            if next_stage == 2:
                self.get_logger().info(
                    '--- Switching to force-position hybrid mode ---'
                )
        self._publish_position_cmd(target)

    def _run_hybrid_stage(self, target, next_stage, label):
        self.target_pos = target
        diff = self._pos_diff_deg(target)
        self.get_logger().info(
            f'[Stage {self.stage}] {label}: '
            f'current={self.current_pos:.4f}, target={target:.4f}, '
            f'diff={diff:.2f} deg'
        )
        if diff < self.POS_THRESHOLD_DEG:
            self.get_logger().info(f'Stage {self.stage} completed')
            self.stage = next_stage
            if next_stage == 4:
                self.get_logger().info('All motion stages completed!')
                self.timer.cancel()
        self._publish_hybrid_cmd(target)

    def _publish_position_cmd(self, pos):
        """Publish position mode command (mode=0)"""
        msg = ArmCtrl()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'arm'
        msg.mode = self.POS_MODE
        msg.label = 0

        ctrl = MotorCtrl()
        ctrl.name = self.JOINT_ID
        ctrl.pos = pos
        ctrl.spd = self.POS_SPD
        ctrl.cur = self.POS_CUR
        msg.ctrl.append(ctrl)

        self.publisher.publish(msg)

    def _publish_hybrid_cmd(self, pos):
        """Publish force-position hybrid mode command (mode=1)"""
        msg = ArmCtrl()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'arm'
        msg.mode = self.HYBRID_MODE
        msg.label = 0

        ctrl = MotorCtrl()
        ctrl.name = self.JOINT_ID
        ctrl.kp = self.HYBRID_KP
        ctrl.kd = self.HYBRID_KD
        ctrl.pos = pos
        ctrl.spd = self.HYBRID_SPD
        ctrl.tor = self.HYBRID_TOR
        msg.ctrl.append(ctrl)

        self.publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = CmdPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
